pkg/controller/job/controller.go

```diff
@@ -255,6 +258,16 @@ func (jm *JobController) enqueueController(obj interface{}) {
 	// all controllers there will still be some replica instability. One way to handle this is
 	// by querying the store for all controllers that this rc overlaps, as well as all
 	// controllers that overlap this rc, and sorting them.
+	// Instead of insert the key directly, check if this key is in the queue already
+	// If it isn't or this key is in the head of the queue, insert the key, otherwise, abort.
+	// If the controller fail and restart, both queue and queuedJob will be lost
+	// everything need to be start over.
+	if jm.queuedJob.Has(key) {
```

Comment:
not exactly I think, we cannot be sure if the head of this queue is in synchronizing stage. If it is, it will still sit in the queue until the synchronization finished but it still need to be enqueue since the synchronization happened this time may be able to apply this update just happened.